### PR TITLE
irmin-bench: bench/tree.exe supports layered store

### PR DESCRIPTION
**Pull request by Ngoguey42**
# (bonus) A series of tests using `bench/tree.exe`
TLDR ; Freeze is extremely sensible to cooperative yields

Test setup: 
- `dune exec -- ./bench/irmin-pack/tree.exe --mode trace --inode-config 32 ~/boostrap_trace.json`.
- Using a layered store.
- 13315 commits total.
- By default, one freeze launched after commit 1664.
- By default, main thread doesn't yield control.
- By default, freeze thread yields control every time it hits a commit.
- After the 13315 commits, we wait for the freeze to end, if still ongoing.

### Variation: Vanilla
As soon as freeze calls Lwt.pause (from `copy to lower`), it is never resumed until the end of the main thread.

Lots of newies.
```
freeze 0 blocked 40.672s (35%), and yielded 1min15s (65%). Total 1min56s.
  Event counts: copy_contents:208847, copy_nodes:1465555, copy_commits:13315, copy_branches:0, adds:1312338, skips:2369230, yields:13316, copy_newies_loops:1
  Longest block 3.454s (during "copy newies (loop)"). Longest yield 1min15s (during "copy to lower"). 114.231 yield per sec.
  Timeline:
    wait for freeze lock took 1.27us (0% of total) and blocked 0.873us (0% of total).
                    misc took 471us (0% of total) and blocked 471us (0% of total).
           copy to lower took 1min20s (69% of total) and blocked 4.661s (11% of total).
      copy to next upper took 0.299us (0% of total) and blocked 0.291us (0% of total).
           copy branches took 3.791us (0% of total) and blocked 3.792us (0% of total).
             flush lower took 131us (0% of total) and blocked 131us (0% of total).
      copy newies (loop) took 36.114s (31% of total) and blocked 35.999s (89% of total).
     wait for batch lock took 1.765us (0% of total) and blocked 0.671us (0% of total).
      copy newies (last) took 65.661us (0% of total) and blocked 65.646us (0% of total).
                    misc took 11.78ms (0% of total) and blocked 11.821ms (0% of total).
                finalize took 358us (0% of total) and blocked 317us (0% of total).
Tezos_log mode on inode config [32, 256]. Results:
895827 "Add" 3.117 sec (4.0%)
54196 "Remove" 1.487 sec (1.9%)
2950152 "Find" 16.010 sec (20.8%)
669692 "Mem" 2.855 sec (3.7%)
439329 "Mem_tree" 0.547 sec (0.7%)
13314 "Checkout" 0.046 sec (0.1%)
59 "Copy" 0.001 sec (0.0%)
13315 "Commit" 52.918 sec (68.7%)
Stats saved to /Users/nico/r/irmin/_results/ad6ffb0c-c7a0-448c-a5e4-67046c2fc0f9/boostrap_trace_timings.json
Total time: 115.013495
```

### Variation: Call `Lwt.pause` after each commit in main thread.
The freeze thread quickly reaches `copy newies (loop)` but it keeps looping in there because the main thread pushes a lot of data.

Lots of newies.
```
freeze 0 blocked 37.963s (32%), and yielded 1min19s (68%). Total 1min57s.
  Event counts: copy_contents:208847, copy_nodes:1465555, copy_commits:13315, copy_branches:0, adds:1312338, skips:2613798, yields:13316, copy_newies_loops:8
  Longest block 2.613s (during "copy newies (loop)"). Longest yield 387ms (during "copy to lower"). 113.432 yield per sec.
  Timeline:
    wait for freeze lock took 0.878us (0% of total) and blocked 0.402us (0% of total).
                    misc took 592us (0% of total) and blocked 592us (0% of total).
           copy to lower took 10.9s (9% of total) and blocked 3.177s (8% of total).
      copy to next upper took 0.411us (0% of total) and blocked 0.397us (0% of total).
           copy branches took 3.991us (0% of total) and blocked 3.992us (0% of total).
             flush lower took 89.446us (0% of total) and blocked 89.86us (0% of total).
      copy newies (loop) took 1min46s (91% of total) and blocked 34.774s (92% of total).
     wait for batch lock took 1.024us (0% of total) and blocked 0.232us (0% of total).
      copy newies (last) took 11.247us (0% of total) and blocked 11.266us (0% of total).
                    misc took 11.696ms (0% of total) and blocked 11.697ms (0% of total).
                finalize took 221us (0% of total) and blocked 220us (0% of total).
Tezos_log mode on inode config [32, 256]. Results:
895827 "Add" 3.318 sec (4.1%)
54196 "Remove" 1.512 sec (1.9%)
2950152 "Find" 15.725 sec (19.6%)
669692 "Mem" 2.458 sec (3.1%)
439329 "Mem_tree" 0.659 sec (0.8%)
13314 "Checkout" 0.047 sec (0.1%)
59 "Copy" 0.000 sec (0.0%)
13315 "Commit" 56.636 sec (70.5%)
Stats saved to /Users/nico/r/irmin/_results/19aa3158-5f9f-41ac-a0fb-80ffbbe46e4e/boostrap_trace_timings.json
Total time: 121.971133
```

### Variation: Call `Lwt_unix.sleep 0.000_000_1` after each commit in main thread.
The freeze thread quickly reaches `copy newies (loop)` but this time it manages to catch up on the main thread after 11 `copy newies` loops.
```
freeze 0 blocked 7.897s (49%), and yielded 8.268s (51%). Total 16.165s.
  Event counts: copy_contents:90383, copy_nodes:514234, copy_commits:3328, copy_branches:0, adds:123077, skips:414837, yields:3329, copy_newies_loops:11
  Longest block 2.262s (during "copy newies (loop)"). Longest yield 352ms (during "copy to lower"). 205.937 yield per sec.
  Timeline:
    wait for freeze lock took 2.27us (0% of total) and blocked 0.45us (0% of total).
                    misc took 357us (0% of total) and blocked 358us (0% of total).
           copy to lower took 6.282s (39% of total) and blocked 3.106s (39% of total).
      copy to next upper took 14.687us (0% of total) and blocked 14.664us (0% of total).
           copy branches took 5.732us (0% of total) and blocked 5.743us (0% of total).
             flush lower took 112us (0% of total) and blocked 112us (0% of total).
      copy newies (loop) took 9.839s (61% of total) and blocked 4.747s (60% of total).
     wait for batch lock took 113us (0% of total) and blocked 112us (0% of total).
      copy newies (last) took 98.457us (0% of total) and blocked 98.444us (0% of total).
                    misc took 25.598ms (0% of total) and blocked 25.598ms (0% of total).
                finalize took 17.579ms (0% of total) and blocked 17.578ms (0% of total).
Tezos_log mode on inode config [32, 256]
895827 "Add" 2.939 sec (4.0%)
54196 "Remove" 1.267 sec (1.7%)
2950152 "Find" 12.640 sec (17.3%)
669692 "Mem" 2.024 sec (2.8%)
439329 "Mem_tree" 0.549 sec (0.8%)
13314 "Checkout" 0.047 sec (0.1%)
59 "Copy" 0.000 sec (0.0%)
13315 "Commit" 53.749 sec (73.4%)
Stats saved to /Users/nico/r/irmin/_results/f35faa99-8c11-4341-a8f6-ac7fa3ad7e62/boostrap_trace_timings.json
Total time: 86.667993
```

### Variation: Call `Lwt_unix.sleep 0.001` after each commit in main thread.
Same, only quicker.
```
freeze 0 blocked 5.968s (65%), and yielded 3.223s (35%). Total 9.192s.
  Event counts: copy_contents:83749, copy_nodes:458641, copy_commits:2444, copy_branches:0, adds:54489, skips:212156, yields:2445, copy_newies_loops:7
  Longest block 2.008s (during "copy newies (loop)"). Longest yield 374ms (during "copy to lower"). 266.004 yield per sec.
  Timeline:
    wait for freeze lock took 1.441us (0% of total) and blocked 0.647us (0% of total).
                    misc took 1.028ms (0% of total) and blocked 1.029ms (0% of total).
           copy to lower took 4.941s (54% of total) and blocked 2.913s (49% of total).
      copy to next upper took 0.37us (0% of total) and blocked 0.279us (0% of total).
           copy branches took 3.505us (0% of total) and blocked 3.506us (0% of total).
             flush lower took 120us (0% of total) and blocked 121us (0% of total).
      copy newies (loop) took 4.228s (46% of total) and blocked 3.032s (51% of total).
     wait for batch lock took 0.554us (0% of total) and blocked 0.265us (0% of total).
      copy newies (last) took 56.936us (0% of total) and blocked 56.938us (0% of total).
                    misc took 20.754ms (0% of total) and blocked 20.754ms (0% of total).
                finalize took 226us (0% of total) and blocked 225us (0% of total).
Tezos_log mode on inode config [32, 256]. Results: ######################] 100%
895827 "Add" 3.477 sec (4.1%)
54196 "Remove" 1.595 sec (1.9%)
2950152 "Find" 18.222 sec (21.7%)
669692 "Mem" 3.064 sec (3.6%)
439329 "Mem_tree" 0.627 sec (0.7%)
13314 "Checkout" 0.126 sec (0.2%)
59 "Copy" 0.001 sec (0.0%)
13315 "Commit" 56.881 sec (67.7%)
Stats saved to /Users/nico/r/irmin/_results/c1256bf4-10ef-40ab-8dd0-220e5c48f2ac/boostrap_trace_timings.json
Total time: 93.481594
```

### Variation: Call `Lwt.pause` after each commit in main thread and remove `Lwt.pause` from freeze thread
The freeze thread totally blocks the main thread
```
freeze 0 blocked 3.092s (100%), and yielded 309us (0%). Total 3.092s.
  Event counts: copy_contents:39355, copy_nodes:220133, copy_commits:1665, copy_branches:0, adds:58, skips:81013, yields:1666, copy_newies_loops:0
  Longest block 2.208s (during "copy to lower"). Longest yield 157us (during "copy to lower"). 538.733 yield per sec.
  Timeline:
    wait for freeze lock took 0.841us (0% of total) and blocked 0.42us (0% of total).
                    misc took 508us (0% of total) and blocked 508us (0% of total).
           copy to lower took 3.058s (99% of total) and blocked 3.057s (99% of total).
      copy to next upper took 0.339us (0% of total) and blocked 0.306us (0% of total).
           copy branches took 4.266us (0% of total) and blocked 4.269us (0% of total).
             flush lower took 150us (0% of total) and blocked 150us (0% of total).
      copy newies (loop) took 1.346us (0% of total) and blocked 1.295us (0% of total).
     wait for batch lock took 0.859us (0% of total) and blocked 0.211us (0% of total).
      copy newies (last) took 60.058us (0% of total) and blocked 60.064us (0% of total).
                    misc took 27.486ms (1% of total) and blocked 27.487ms (1% of total).
                finalize took 6.389ms (0% of total) and blocked 6.388ms (0% of total).
Tezos_log mode on inode config [32, 256]
895827 "Add" 3.087 sec (3.8%)
54196 "Remove" 1.661 sec (2.0%)
2950152 "Find" 20.873 sec (25.7%)
669692 "Mem" 2.462 sec (3.0%)
439329 "Mem_tree" 0.588 sec (0.7%)
13314 "Checkout" 0.050 sec (0.1%)
59 "Copy" 0.001 sec (0.0%)
13315 "Commit" 52.512 sec (64.6%)
Stats saved to /Users/nico/r/irmin/_results/a247c732-f596-41cf-93ab-0c2ada9bd96a/boostrap_trace_timings.json
Total time: 85.884298
```

### CODE REVIEW

File: bench/irmin-pack/tree.ml [hunk 1]
```diff
@@ -16,10 +16,39 @@ type op =
...
+  ncommits_trace : int;
+  depth : int;
+  nchain_trees : int;
+  width : int;
+  nlarge_trees : int;
+  root : string;
+  flatten : bool;
+  inode_config : [ `Entries_32 | `Entries_2 ];
+  store_type : [ `Pack | `Pack_layered ];
+  freeze_commit : int;
```
Comment (icristescu) [inline bench/irmin-pack/tree.ml:30]: maybe you can combine the two: it does not make sense to have a `freeze_commit` if the store type is not layered.

File: bench/irmin-pack/tree.ml [hunk 2]
```diff
@@ -16,10 +16,39 @@ type op =
...
+  commit_data_file : string;
+  results_dir : string;
+}
+
+module type STORE = sig
+  include Irmin.S with type key = string list and type contents = string
+
+  type on_commit := int -> Hash.t -> unit Lwt.t
```
Comment (icristescu) [inline bench/irmin-pack/tree.ml:38]: nice trick :)

File: bench/irmin-pack/tree.ml [hunk 3]
```diff
@@ -364,39 +371,43 @@ struct
...
-        "Large trees mode on inode config %a: %d commits, each consisting of \
-         %d large trees of %d entries\n\
+        "Large trees mode on inode config %a, %a: %d commits, each consisting \
+         of %d large trees of %d entries\n\
+         %a@\n\
          %a"
-        pp_inode_config config.inode_config config.ncommits config.nlarge_trees
-        config.width Benchmark.pp_results result
+        pp_inode_config config.inode_config pp_store_type config.store_type
+        config.ncommits config.nlarge_trees config.width repo_pp ()
```
Comment (icristescu) [inline bench/irmin-pack/tree.ml:389]: you can use `%t` to avoid the extra unit argument for `repo_pp` (as in a previous PR)

File: bench/irmin-pack/tree.ml [hunk 4]
```diff
@@ -405,43 +416,112 @@ struct
...
+          let c = Option.get c in
+          Store.freeze repo ~max:[ c ] ~min_upper:[ c ]
+        else Lwt.return_unit
+      in
+      let* () =
+        if i = ncommits - 1 then Store.PrivateLayer.wait_for_freeze repo
+        else Lwt.return_unit
+      in
+      Lwt.pause ()
+      (* Lwt_unix.sleep 0.001 *)
```
Comment (icristescu) [inline bench/irmin-pack/tree.ml:478]: you forgot to remove this comments;
if you think benchmarking this type of behaviour for the main thread is pertinent maybe we should add it explicitly? I'm on a fence about adding pause in the main thread, as in Tezos, we won't do this. But maybe it's useful to detect blocking behaviour in the freeze thread?

**Review by icristescu — APPROVED** [review]
Thanks, LGTM

File: bench/irmin-pack/tree.ml (hunk 1 above)
Comment (Ngoguey42) [reply bench/irmin-pack/tree.ml:30]: This data-structure already mixes things (e.g. `flatten` and `ncommits_trace` only apply to trace mode). Adding `freeze_commit` here is consistent with what we have been doing so far in this file.

We could clean this in a latter patch

File: bench/irmin-pack/tree.ml (hunk 3 above)
Comment (Ngoguey42) [reply bench/irmin-pack/tree.ml:389]: Thanks, I forgot about this one

File: bench/irmin-pack/tree.ml (hunk 4 above)
Comment (Ngoguey42) [reply bench/irmin-pack/tree.ml:478]: As I mentioned in Wednesday's meeting, I highly doubt we will be able to leave the main thread's yields to chance while keeping an `Lwt.async` freeze. If we want a robust behaviour, we will at least have to change something in the way freeze works. Detecting blocking behaviours in the freeze thread would be a great start.

I don't think this parameter is pertinent yet, it will depend on the solutions we think about.

I indeed forgot to remove these comments, I will replace these with a small explanation.
